bin/add_dois.py

```diff
@@ -0,0 +1,141 @@
+        has_doi = False
+        if frontmatter is not None:
+            old_doi_text = ""
+            
+            for doi in frontmatter.findall('doi'):
+                has_doi = True
+                old_doi_text = doi.text
+
+                if (not has_doi or args.force): # need to assign DOI
+                    new_doi_text = args.prefix + collection_id + "-" + volume_id
+                    doi = ""
+                    
+                    if (args.force and has_doi):
+                        print(f'Overwritting existing booktitle DOI {old_doi_text} with {new_doi_text}', file=sys.stderr)
+                        doi = frontmatter.find('doi')
+                    else:
+                        print(f'Writing booktitle DOI {new_doi_text}', file=sys.stderr)
+                        doi = ET.Element('doi')
+                        
+                        doi.text = new_doi_text
+                        
+                        # Set tails to maintain proper indentation
+                        frontmatter[-1].tail += '  '
+                        doi.tail = '\n    ' # newline and two levels of indent
+                        frontmatter.append(doi)
+                        n += 1
+                        
+        if (has_doi and not args.force):
+            print(f'Cowardly refusing to overwrite existing booktitle DOI.  Use --force', file=sys.stderr)
+            
+        has_doi = False
+        # Iterate through all papers
+        for paper in volume.findall('paper'):
+            old_doi_text = ""
+
+            # check if DOI exists
+            for doi in paper.findall('doi'):
+                has_doi = True
+                old_doi_text = doi.text
+
+            if (not has_doi or args.force): # need to assign DOIs
+                new_doi_text = args.prefix + build_anthology_id(collection_id, volume_id, paper.get('id'))
+                doi = ""
+
+                if (args.force and has_doi):
+                    print(f'Overwritting existing DOI {old_doi_text} with {new_doi_text}', file=sys.stderr)
+                    doi = paper.find('doi')
+                else:
+                    print(f'Writing DOI {new_doi_text}', file=sys.stderr)
+                    doi = ET.Element('doi')
+
+                doi.text = new_doi_text
+
+                # Set tails to maintain proper indentation
+                paper[-1].tail += '  '
+                doi.tail = '\n    '  # newline and two levels of indent
+                paper.append(doi)
+                n += 1
+        if (has_doi and not args.force):
+            print(f'Cowardly refusing to overwrite existing paper DOIs.  Use --force', file=sys.stderr)
+
+        tree.write(xml_file, encoding="UTF-8", xml_declaration=True)
+        print(f'-> added {n} DOIs to to the XML for collection {collection_id}', file=sys.stderr)
+
+    else:
+        print(f'-> FATAL: volume {volume} not found in the Anthology', file=sys.stderr)
+        sys.exit(1)
+
```

Comment:
There's a lot of redundancy in these checks for front matter and papers—can we factor that piece out?